apps/rush-lib/src/logic/ProjectTaskSelector.ts

```diff
@@ -70,84 +70,94 @@ export class ProjectTaskSelector {
+          selectedDependencies.add(dependency);
+        } else {
+          // This project is not part of the set, but some of its dependencies could be.
+          for (const indirectDependency of getSelectedDependencies(dependency)) {
+            selectedDependencies.add(indirectDependency);
+          }
+        }
+      }
 
-  private _registerProjectPhaseTask(
-    project: RushConfigurationProject,
-    phase: IPhase,
-    taskCollection: TaskCollection
-  ): string {
-    const taskName: string = this._getPhaseDisplayNameForProject(phase, project);
-    if (taskCollection.hasTask(taskName)) {
-      return taskName;
+      return selectedDependencies;
     }
 
-    const customParameterValues: string[] = this._getCustomParameterValuesForPhase(phase);
-    const commandToRun: string | undefined = ProjectTaskSelector.getScriptToRun(
-      project,
-      phase.name,
-      customParameterValues
-    );
-    if (commandToRun === undefined && !phase.ignoreMissingScript) {
-      throw new Error(
-        `The project [${project.packageName}] does not define a '${phase.name}' command in the 'scripts' section of its package.json`
-      );
-    }
+    // Register all tasks
+    // This currently does not correctly handle the scenario in which a phase is skipped.
+    // If that happens, the dependency graph will have an error due to a missing task.
+    for (const phaseName of this._phasesToRun) {
+      const phase: IPhase = this._getPhaseByName(phaseName);
+      for (const project of projects) {
+        const taskName: string = this._getPhaseDisplayNameForProject(phase, project);
 
-    taskCollection.addTask(
-      new ProjectTaskRunner({
-        rushProject: project,
-        taskName,
-        rushConfiguration: this._options.rushConfiguration,
-        buildCacheConfiguration: this._options.buildCacheConfiguration,
-        commandToRun: commandToRun || '',
-        isIncrementalBuildAllowed: this._options.isIncrementalBuildAllowed,
-        projectChangeAnalyzer: this._projectChangeAnalyzer,
-        phase
-      })
-    );
-
-    const dependencyTasks: Set<string> = new Set();
-    if (phase.dependencies?.self) {
-      for (const dependencyPhaseName of phase.dependencies.self) {
-        const dependencyPhase: IPhase = this._getPhaseByName(dependencyPhaseName);
-        const dependencyTaskName: string = this._registerProjectPhaseTask(
+        const customParameterValues: string[] = this._getCustomParameterValuesForPhase(phase);
+        const commandToRun: string | undefined = ProjectTaskSelector.getScriptToRun(
           project,
-          dependencyPhase,
-          taskCollection
+          phase.name,
+          customParameterValues
         );
+        if (commandToRun === undefined && !phase.ignoreMissingScript) {
+          throw new Error(
+            `The project [${project.packageName}] does not define a '${phase.name}' command in the 'scripts' section of its package.json`
```

Comment:
The `[ ... ]` square brackets are a bit odd here. I think we normally just put the package name in other locations.

Reply:
```suggestion
            `The project '${project.packageName}' does not define a '${phase.name}' command in the 'scripts' section of its package.json`
```